[autoscaler] Refactor locks behavior in local node provider

jyakaranda opened the pull request:
## Why are these changes needed?

1. keep locks private in ClusterState;
2. add a new function `get_for_update` to update workers outside ClusterState.

to prevent issues where only one lock might be acquired or locks are acquired in 
inconsistent order.

per comment from: https://github.com/ray-project/ray/pull/49909#issuecomment-2797878328

prerequisite PR: https://github.com/ray-project/ray/pull/49909

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

rueian commented on python/ray/autoscaler/_private/local/node_provider.py:
```diff
@@ -246,15 +255,15 @@ def internal_ip(self, node_id):
 
     def set_node_tags(self, node_id, tags):
-        with self.state.file_lock:
+        with self.state.combined_lock:
```
Hi @jyakaranda, Thanks for the follow-up!

When I mentioned `encapsulating the lock and file_lock entirely` previously, I didn't mean combining the two locks. What I meant was that we should keep those locks private and should not access them from the outside LocalNodeProvider. To achieve that, we would need to merge `state.get()` and `state.put()`.

Sorry for the confusion.

jyakaranda replied:
okay, keeping locks private is a good idea, and it seems straightforward to using single updated `put` in `set_node_tags` and `terminate_node`.
But I got a question about `create_node`, merge `state.get` and `state.put` might be insufficient here imho, since it updates multiple nodes in an atomic way. Shall we introduce a new function in `ClusterState` and move the creating node logic into it?
I'm not sure if I understand it right and welcome your feedbacks

rueian replied:
I think a new `get_for_update` function is good. It can take another function as an argument, and that function will mutate the fresh state and return a new one.

jyakaranda replied:
ok, I updated the code, PTAL~

jyakaranda committed:
[autoscaler] Refactor locks behavior in local node provider

1. keep locks private in ClusterState;
2. add a new function `get_for_update` to update workers outside
   ClusterState.

Signed-off-by: jyakaranda <zhangheng1261@gmail.com>

rueian commented on python/ray/autoscaler/_private/local/node_provider.py:
```diff
@@ -117,6 +117,30 @@ def put(self, worker_id, info):
+        with self._lock:
+            with self._file_lock:
+                workers = json.loads(open(self.save_path).read())
+                result = worker_update_fn(workers)
```
`worker_update_fn` looks confusing. Can we use just `update_fn`?

jyakaranda replied:
sure, updated

rueian commented on python/ray/autoscaler/_private/local/node_provider.py:
```diff
@@ -117,6 +117,30 @@ def put(self, worker_id, info):
+            with self._file_lock:
+                workers = json.loads(open(self.save_path).read())
+                result = worker_update_fn(workers)
+                if result is not None:
```
Can't we save whatever the `update_fn` returned?

jyakaranda replied:
it's fine to save whatever the `update_fn` return in current usage, I removed this check

rueian commented on python/ray/autoscaler/_private/local/node_provider.py:
```diff
@@ -98,16 +98,16 @@ def __init__(self, lock_path, save_path, provider_config):
                 workers = json.loads(open(self.save_path).read())
                 return workers
 
     def put(self, worker_id, info):
```
Can we remove the `put` method?

rueian replied:
Are there any existing tests that cover the new get_for_update method? Otherwise, we need to add one.

jyakaranda replied:
yes, we can remove `put`.
there is a `OnPremCoordinatorServerTest` that would cover code changes
https://github.com/ray-project/ray/blob/0a0aad6b00dabc25134d75d7cbdc34ebc3ecd8c1/python/ray/tests/test_coordinator_server.py#L31

jyakaranda committed:
Update according to rueian's comments

1. remove `put` from ClusterState;
2. update get_for_update.

Signed-off-by: jyakaranda <zhangheng1261@gmail.com>
